readthedocs/donate/models.py

```diff
@@ -50,9 +59,48 @@ def __str__(self):
 
     def as_dict(self):
         "A dict respresentation of this for JSON encoding"
+        image_url = reverse(
+            'donate_view_proxy',
+            kwargs={'promo_id': self.pk, 'hash': random.randint(0, 10000000)}
+        )
+        link_url = reverse(
+            'donate_click_proxy',
+            kwargs={'promo_id': self.pk}
+        )
         return {
             'id': self.analytics_id,
             'text': self.text,
-            'link': self.link,
-            'image': self.image,
+            'link': link_url,
+            'image': image_url,
         }
+
+    def incr(self, type):
+        assert type in ['offers', 'views', 'clicks']
+        day = get_ad_day()
+        impression, _ = self.impressions.get_or_create(date=day)
+        setattr(impression, type, models.F(type) + 1)
+        impression.save()
+
+    def incr_redis(self, type):
+        assert type in ['offers', 'views', 'clicks']
+        day = get_ad_day()
+        redis = Redis.from_url(settings.BROKER_URL)
+        redis.incr('{slug}-{year}-{month}-{day}-{type}'.format(
```

Comment:
We can then look back into using Redis once we know the query patterns and indexes we want. Can just use SQL's flexibility for now, and build Redis indexes later if we need them.